source/GameInterface/Services/BesiegerCamps/Extensions/BesiegerCampExtensions.cs

```diff
@@ -0,0 +1,50 @@
+﻿using GameInterface.Services.ObjectManager;
+using Serilog;
+using System;
+using System.Reflection;
+using TaleWorlds.CampaignSystem.Siege;
+
+namespace GameInterface.Services.BesiegerCamps.Extensions
+{
+    public static class BesiegerCampExtensions
+    {
+        public static string TryGetId(object value, ILogger logger)
+        {
+            if (value == null) return null;
+
+            // temp fix for SiegeStrategy not being registered
+            if (value is SiegeStrategy siegeStrategy)
+            {
+                return siegeStrategy.StringId;
+            }
+
+            if (!ContainerProvider.TryResolve<IObjectManager>(out var objectManager))
+            {
+                logger.Error("Unable to resolve {type}", typeof(IObjectManager).FullName);
+                return null;
+            }
+
+            if (!objectManager.TryGetId(value, out string typeId))
+            {
+                logger.Error("Unable to get ID for instance of type {type}", value.GetType().Name);
+                return null;
+            }
+
+            return typeId;
+        }
+
+        // quick and dirty way to pass the type as refference isntead of type arg
+        public static bool TryGetObject(this IObjectManager src, string id, Type type, out object obj)
```

Comment:
It's a version of ObjectManager's `TryGetObject<T>` with the type passed as reference instead of a type arg.

Its used on the `HandleDataChanged` methods where I receive an generic message with the name of the property being changed and the new value, I only know the type of the property being changed at runtime so `TryGetObject<T>` doesn't work.

I reservedly made this extension method because I didn't wanted to change anything outside the service of this issue, but adding this overload to `ObjectManager` would fill this need

```
    public bool TryGetObject(string id, Type type, out object obj)
    {
        obj = default;

        if (string.IsNullOrEmpty(id)) return false;

        if (RegistryMap.TryGetValue(type, out IRegistry registry))
        {
            return registry.TryGetValue(id, out obj);
        }

        /// Default object manager <see cref="MBObjectManager"/> requires type to be <see cref="MBObjectBase"/>
        return defaultObjectManager.TryGetObject(id, out obj);
    }
```
I could also refactor out my need of it by creating a handler for each property type, but that would add a lot of boilerplate

Reply:
Done